domains/runtime/evm/src/lib.rs

```diff
@@ -817,6 +827,75 @@ mod benches {
     );
 }
 
+/// Custom pre_dispatch for extrinsic verification.
+/// Most of the logic is same as `pre_dispatch_self_contained` except
+/// - we use `validate_self_contained` instead `pre_dispatch_self_contained`
+///   since the nonce is not incremented in `pre_dispatch_self_contained`
+/// - Manually track the account nonce to check either Stale or Future nonce.
+fn pre_dispatch_evm_transaction(
+    account_id: H160,
+    call: RuntimeCall,
+    dispatch_info: &DispatchInfoOf<RuntimeCall>,
+    len: usize,
+) -> Result<(), TransactionValidityError> {
+    match call {
+        RuntimeCall::Ethereum(call) => {
+            // Withdraw the consensus chain storage fee from the caller and record
+            // it in the `BlockFees`
+            let consensus_storage_fee =
+                BlockFees::consensus_chain_byte_fee() * Balance::from(len as u32);
+            match <InnerEVMCurrencyAdapter as pallet_evm::OnChargeEVMTransaction<Runtime>>::withdraw_fee(
+                &account_id,
+                consensus_storage_fee.into(),
+            ) {
+                Ok(None) => {}
+                Ok(Some(paid_consensus_storage_fee)) => {
+                    BlockFees::note_consensus_storage_fee(paid_consensus_storage_fee.peek())
+                }
+                Err(_) => return Err(InvalidTransaction::Payment.into()),
+            }
+
+            if let Some(transaction_validity) =
+                call.validate_self_contained(&account_id, dispatch_info, len)
+            {
+                transaction_validity.map(|_| ())?;
+
+                if let Call::transact { transaction } = call {
+                    CheckWeight::<Runtime>::do_pre_dispatch(dispatch_info, len)?;
+
+                    let transaction_data: TransactionData = (&transaction).into();
+                    let transaction_nonce = transaction_data.nonce;
+                    // if this the first transaction from this sender, then use the
+                    // transaction nonce as first nonce.
```

Comment:
That would not work when we want to send multiple successive transactions in successive bundle within the same consensus block.

We assume the transactions are included in the nonce order due to how tx_pool gives the ready transactions. 
If the first bundle contains nonce from 0, 1, 2, 3
Second bundle 4, 5, 6

If we take the tracked nonce 0 like you have suggested, then second bundle will never include 4, 5, 6 since they will be considered Future and rejects them

The caveat with this approach is that custom pre_dispatch should not be abused for other cases. In our case, as long as we use to bundle transactions coming from tx_pool in order, we should be able to include all the transactions in all successive bunldes within a given consensus block.

Maybe I'll add a comment here and note for future reference.

Reply:
>That would not work when we want to send multiple successive transactions in successive bundle within the same consensus block.

This is not supported before and can't be simply supported by the changes in this PR either.

Although, we are using the same runtime API instance for checking `check_extrinsics_and_do_pre_dispatch` for all tx in the same bundle, for different bundles we are using different runtime API instances which don't know about tx in other bundles, and all bundles are verified against the state of the last domain block, if the nonce is 0 in the last block then all bundle need to include tx with nonce start from 0, bundle with 4,5,6 will be considered as invalid bundle.

Also, that doesn't mean all bundles will have to include the exact set of tx, a bundle may include 0,1,2 while another may include 0,1,2,3,4 due to the tx range is changing every slot, and tx from other sender may not fall in the range so make room for this sender.

>The caveat with this approach is that custom pre_dispatch should not be abused for other cases. In our case, as long as we use to bundle transactions coming from tx_pool in order, we should be able to include all the transactions in all successive bunldes within a given consensus block.

If we want to support this scenario, we need to think more about the security aspect, i.e how we prevent the malicious operator from abusing this to intentionally skip tx with nonce 0,1,2,3 and produce bundles with useless tx with nonce 4,5,6...

Reply:
> This is not supported before and can't be simply supported by the changes in this PR either.

I disagree. This should be supported with the current PR changes. Maybe I'm missing something. If so, please elaborate.

> Although, we are using the same runtime API instance for checking check_extrinsics_and_do_pre_dispatch for all tx in the same bundle, for different bundles we are using different runtime API instances which don't know about tx in other bundles, and all bundles are verified against the state of the last domain block, if the nonce is 0 in the last block then all bundle need to include tx with nonce start from 0, bundle with 4,5,6 will be considered as invalid bundle.

yes that is why I have mentioned, the caveat of not abusing it.
Sure a malicious operator with a different reference implementation may abuse this to include the transactions out of the order. I do not have a good solutions yet for this case with purely runtime changes without tracking it on client side. Even if we track it on client side, there is not garuntee that malicious operator will not change it.

If the malicious operator skips txns 0,1,2 and include 3, 4, 5. There can be two scenarios
- Any bundles with 0,  1, 2 is already included, then this scenario should be fine 
- Any bundles without 0, 1, 2, then malicious operator will be slashed for producing the invalid bundle

> Also, that doesn't mean all bundles will have to include the exact same set of tx, a bundle may include 0,1,2 while another may include 0,1,2,3,4 due to the tx range is changing every slot, and tx from other sender may not fall in the range so make room for this sender.

Yeah so why would this be a problem is something I'm not understanding here. This change is self contained from the perspective of single operator who is pushing the bundles in same consensus block. They will not include same transactions in successive bundles. If a different operator submits a bundle with more transactions since they were elected for a different slot, then sure they will include more txns given they are not seeing other transactions from the network.

Reply:
> If the malicious operator skips txns 0,1,2 and include 3, 4, 5. There can be two scenarios

I think this is what missing.
We should not use same check_pre_dispatch while proposing bundles and while validating bundles from the consensus block.

While for the first one, we can use the current PR changes but for the validating the bundles, it should use the same runtime instance for all the bundles from consensus block.
This should essentially handle the second case where if the malicious operator did indeed skip the transactions, then bundle will become invalid and there by slashing the operator

Reply:
Looking into this further. This is quiet unfortunate the way we designed the pre_dispatch check before we submit a bundle. This does not seem efficient at all when there are a lot of transactions that can all go on within single consensus block.Instead our design is limiting to just filling up first few and potentially looking to fill other transactions when clearly there are txns in pool that are ready. Seems like a design issue.


I'll rest the change in this PR but I'm going to open an issue regarding this.

Reply:
What this PR implemented for this feature is the most simple happy path.

For the malicious operator, it is okay and feasible to use the same runtime instance for verifying all bundles produced by the same operator, but this is just how we detect the fraud, what tricky is how we prove it in fraud proof.

Right now, the illegal bundle FP is targeting a single bundle so it only needs to carry 1 bundle (whose size is bounded by the bundle limit) and the execution proof of doing `pre_dispatch` from the first tx to the illegal tx. To support this feature we need to change the FP to target all bundles produced by the same operator, i.e. if the illegal tx is in the last bundle then we will need all the previous bundles and the execution proof of executing all tx in these bundles, the size is overwhelming for a single FP.

Even without considering the malicious operator, for an honest operator produced bundle1 with 0,1,2 and bundle2 with 3,4,5, it is still possible that
- bundle2 is included in the consensus block while bundle1 is not
- bundles are arrived and included in the block out of order

From my understanding of the de-duplicate & shuffle process, these will still cause the extrinsic ordering issue, though it can potentially fixed with `require` & `provvide` tag in tx pool.

Overall, I'm not against with this feature (I did also considered it briefly previously) but it is certainly not a low hanging fruit that can be implemented easily.